moveit_core/trajectory_processing/include/moveit/trajectory_processing/time_optimal_trajectory_generation.h

```diff
@@ -0,0 +1,179 @@
+public:
+  /// @brief Generates a time-optimal trajectory
+  Trajectory(const Path& path, const Eigen::VectorXd& max_velocity, const Eigen::VectorXd& max_acceleration,
+             double time_step = 0.001);
+
+  ~Trajectory(void);
+
+  /** @brief Call this method after constructing the object to make sure the
+     trajectory generation succeeded without errors. If this method returns
+     false, all other methods have undefined behavior. **/
+  bool isValid() const;
+
+  /// @brief Returns the optimal duration of the trajectory
+  double getDuration() const;
+
+  /** @brief Return the position/configuration vector for a given point in time
+   */
+  Eigen::VectorXd getPosition(double time) const;
+  /** @brief Return the velocity vector for a given point in time */
+  Eigen::VectorXd getVelocity(double time) const;
+  /** @brief Return the acceleration vector for a given point in time */
+  Eigen::VectorXd getAcceleration(double time) const;
+
+private:
+  struct TrajectoryStep
+  {
+    TrajectoryStep()
+    {
+    }
+    TrajectoryStep(double path_pos, double path_vel) : path_pos_(path_pos), path_vel_(path_vel)
+    {
+    }
+    double path_pos_;
+    double path_vel_;
+    double time_;
+  };
+
+  bool getNextSwitchingPoint(double path_pos, TrajectoryStep& next_switching_point, double& before_acceleration,
+                             double& after_acceleration);
+  bool getNextAccelerationSwitchingPoint(double path_pos, TrajectoryStep& next_switching_point,
+                                         double& before_acceleration, double& after_acceleration);
+  bool getNextVelocitySwitchingPoint(double path_pos, TrajectoryStep& next_switching_point, double& before_acceleration,
+                                     double& after_acceleration);
+  bool integrateForward(std::list<TrajectoryStep>& trajectory, double acceleration);
+  void integrateBackward(std::list<TrajectoryStep>& start_trajectory, double path_pos, double path_vel,
+                         double acceleration);
+  double getMinMaxPathAcceleration(double path_position, double path_velocity, bool max);
+  double getMinMaxPhaseSlope(double path_position, double path_velocity, bool max);
+  double getAccelerationMaxPathVelocity(double path_pos) const;
+  double getVelocityMaxPathVelocity(double path_pos) const;
+  double getAccelerationMaxPathVelocityDeriv(double path_pos);
+  double getVelocityMaxPathVelocityDeriv(double path_pos);
+
+  std::list<TrajectoryStep>::const_iterator getTrajectorySegment(double time) const;
+
+  Path path_;
+  Eigen::VectorXd max_velocity_;
+  Eigen::VectorXd max_acceleration_;
+  unsigned int joint_num_;
+  bool valid_;
+  std::list<TrajectoryStep> trajectory_;
+  std::list<TrajectoryStep> end_trajectory_;  // non-empty only if the trajectory generation failed.
+
+  static constexpr double EPS = 0.000001;
+  const double time_step_;
+
+  mutable double cached_time_;
+  mutable std::list<TrajectoryStep>::const_iterator cached_trajectory_segment_;
```

Comment:
why are these mutable? this should be avoided

Reply:
These are mutable to allow caching results to speed up the const function `getTrajectorySegment()`. It's probably better than removing this feature or making the getter function non-const.

Reply:
what if you have two instances of this class in different parts of the code base and sharing this memory would be a bug? this is what member variables are for

Reply:
GStyle Guide:

> The mutable keyword is allowed but is unsafe when used with threads, so thread safety should be carefully considered first.

Reply:
These are member variables of Trajectory and two instances of Trajectory never share any memory.
The iterator `cached_trajectory_segment_` is generated from the private member list `trajectory_` (which never changes after the constructor) and both mutables are only changed inside `getTrajectorySegment()`. I don't know any way this could cause trouble. Do you see a better alternative for removing the mutable specifier than making all 'getter()' functions non-const?

Reply:
Ah I think I see what you're saying. I'd still like to get @rhaschke 's opinion on this.

Reply:
Cache variables are typical candidates for the mutable keyword. I fully agree with Henning, that these variables are save here as long as the Trajectory instance isn't accessed in parallel from multiple threads. But the chances for this are really low, aren't they?

Reply:
Correct, this doesn't happen at all since each Trajectory instance only lives inside `computeTimeStamps()` which doesn't create multiple threads.

Reply:
+1